### DISCUSSION TITLE
OAuth session termination on logout

### DISCUSSION THREAD

**Discussion by alexanderwolz** [directus/directus#12132] [Archive]
Hi there,

I would like to discuss a feature request "enabling oauth/openidc session termination upon sign-out". See also issue https://github.com/directus/directus/issues/12127

Does it make sense to terminate the actual oauth session if user actively signs out from Directus app?
I believe yes, because the user actively ends its session by clicking "sign-out" and assumes all tokens to be revoked. Directus then would re-trigger the authentication flow of the IAM server.

Currently, upon re-login, the auth flow is not triggered but the refresh token of the previously logged-in user is used.

Besides setting ```AUTH_PROVIDER_PARAMS__PROMPT: "login"``` for triggering re-authentication may work in order to switch users but is not a nice user experience (also there seems to be a bug in the Directus session handling). If a user logged out, he should be asked for credentials again at it's OAuth provider. I believe the cleanest way would be calling the end-session-endpoint of the IAM server e.g. 

```
curl --request POST \
  --url https://myauthserver.com/realms/myrealm/protocol/openid-connect/logout \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data client_id=myClient \
  --data client_secret=mySecret \
  --data refresh_token=ey_MyRefreshToken
```

What do you think?

**Comment by aidenfoxx** [comment]
I would just like to mention that it was an intentional design decision to not revoke oauth access on logout to allow for more fluid authentication in general e.g. the customer doesn't have to re-trigger the authorization prompt every time they log in via SSO unless the "prompt" param is set or the user revokes access on the provider side.

With that being said, this would have to be enabled as an optional flag if we want to keep the current default SSO experience.

**Reply by fabiobeoni** [reply]
Configuring an `AUTH_<PROVIDER>_LOGOUT_URL` Directus could switch the default logout behavior to the one proposed.

**Comment by mlbiche** [comment]
I confirm that I have the same behaviour when plugin Directus with Keycloak 19. Using the `prompt=login` query parameters indeed leads the user to the login page. But the login page only works if the user uses the same credentials. If I try to log in with another user, it fails.

**Reply by aidenfoxx** [reply]
Does your configuration in Directus allow auto user registration? Otherwise you have to ensure there is a user already configured in Directus that is compatible with your Keycloak  login.

**Reply by mlbiche** [reply]
Sorry, I was talking about the Keycloak login page which expect the authentication to be done with the same user.

**Reply by aidenfoxx** [reply]
I'm not sure what to suggest, unless it's a bug on the Keycloak side. Their docs says that for "login" prompt "SSO will be ignored and the Keycloak login page will be always shown, even if the user is already authenticated". https://www.keycloak.org/docs/latest/securing_apps/#_params_forwarding

**Reply by aidenfoxx** [reply]
That being said, contributions to improve functionality are always encouraged. The ability to revoke access on logout would be a welcome option, though might be tricky since we currently don't store provider access tokens after authentication.

**Reply by mlbiche** [reply]
Ok, I may have missed something while testing. What I was talking about is when the user is already authenticated and, he accesses to the login page because of the `prompt=login` query param, the login page is already set to its account, as you can see in the red rectangle :

<img width="778" alt="Capture d’écran 2022-10-26 à 10 22 38" src="https://user-images.githubusercontent.com/33522659/197974234-a07f4a77-03a9-4480-8101-7f2380e631b3.png">

When submitting other credentials, the authentication fails, even if credentials are correct. The solution to authenticate with another account is to click on the blue square and arrow at the left of the user email, which leads to a normal login. But it is not really user-friendly.

But anyway, it is Keycloak-related then 👌
